eZ/Publish/Core/Repository/ContentService.php

```diff
@@ -1742,6 +1742,21 @@ protected function copyTranslationsFromPublishedVersion(APIVersionInfo $versionI
         $this->internalUpdateContent($versionInfo, $updateStruct);
     }
 
+    protected function isHashEqual($fieldType, $newValue, $fieldValue): bool
+    {
+        $newHash = $fieldType->toHash($newValue);
+        $currentHash = $fieldType->toHash($fieldValue);
+        if ($newHash === $currentHash) {
+            return true;
+        } elseif ($fieldType instanceof \eZ\Publish\Core\FieldType\Image\Type) {
+            $imageHashDiff = array_diff($newHash, $currentHash);
+
+            return count($imageHashDiff) === 1 && !empty($imageHashDiff['imageId']);
+        }
```

Comment:
This is not SOLID. It violates open-closed principle. Imagine you'll have another case for different FT in the future. Another `if`? What if the case is for 3rd party FT?
It's a responsibility of FT ValueObject (or any VO for that matter) to know it's equality conditions. You need to delegate this operation to FT specific implementation. Of course due to BC this method comparing `$newHash === $currentHash` should stay as a fallback.

It can be implemented by adding new interface e.g. `eZ\Publish\SPI\FieldType\Value\Comparable` with method `equals(Value)`
Then you can go with `if ($newValue instanceof Comparable)` ....